packages/middleware/rate-limiter/src/sliding-window.ts

```diff
@@ -0,0 +1,125 @@
+import { Rate, RateLiteral, parseRate } from './rate'
+import { Slot, SlotProvider } from './slot'
+import { types } from '@mondrian-framework/model'
+import { functions } from '@mondrian-framework/module'
+
+//inspired by https://blog.cloudflare.com/counting-things-a-lot-of-different-things/
+
+/**
+ * Rate limit at `rateLimit` in the `samplingPeriod`.
+ * Example of maximum 10 request every 2 minutes:
+ * ```typescript
+ * new SlidingWindow({ rateLimit: 10, samplingPeriod: "2 minutes" })
+ * ```
+ */
+export class SlidingWindowProvider {
+  private readonly slidingWindows: Map<string, SlidingWindow> = new Map()
+  private readonly rate: Rate
+  private readonly slotProvider: SlotProvider
+
+  constructor({ rate, slotProvider }: { rate: Rate | RateLiteral; slotProvider: SlotProvider }) {
```

Comment:
Also I think it should only accept a `Rate`, it's not this class responsibility to parse a `RateLiteral` into a `Rate`; it should be the outside caller that has to provide it an already valid `Rate`